pkg/cmd/admin/router/router.go

```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
+	}
+	env.Add(secretEnv)
+	if len(defaultCert) > 0 {
+		if cfg.SecretsAsEnv {
+			env.Add(app.Environment{"DEFAULT_CERTIFICATE": string(defaultCert)})
+		} else {
+			// TODO: make --credentials create secrets and bypass service account
+			env.Add(app.Environment{"DEFAULT_CERTIFICATE_PATH": defaultCertificatePath})
 		}
+	}
 
-		if len(secrets) != 0 {
-			serviceAccount, err := kClient.ServiceAccounts(namespace).Get(cfg.ServiceAccount)
-			if err != nil {
-				return fmt.Errorf("error looking up service account %s: %v",
-					cfg.ServiceAccount, err)
-			}
-
-			for _, secret := range secrets {
-				objects = append(objects, secret)
+	updatePercent := int(-25)
 
-				serviceAccount.Secrets = append(serviceAccount.Secrets,
-					kapi.ObjectReference{Name: secret.Name})
-			}
+	secrets, volumes, mounts, err := generateSecretsConfig(cfg, kClient,
+		namespace, defaultCert)
+	if err != nil {
+		return fmt.Errorf("router could not be created: %v", err)
+	}
 
-			_, err = kClient.ServiceAccounts(namespace).Update(serviceAccount)
-			if err != nil {
-				return fmt.Errorf("error adding secret key to service account %s: %v",
-					cfg.ServiceAccount, err)
-			}
-		}
+	livenessProbe := generateLivenessProbeConfig(cfg, ports)
+	readinessProbe := generateReadinessProbeConfig(cfg, ports)
 
-		objects = app.AddServices(objects, true)
-		// TODO: label all created objects with the same label - router=<name>
-		list := &kapi.List{Items: objects}
+	exposedPorts := make([]kapi.ContainerPort, len(ports))
+	copy(exposedPorts, ports)
+	for i := range exposedPorts {
+		exposedPorts[i].HostPort = 0
+	}
+	containers := []kapi.Container{
+		{
+			Name:            "router",
+			Image:           image,
+			Ports:           exposedPorts,
+			Env:             env.List(),
+			LivenessProbe:   livenessProbe,
+			ReadinessProbe:  readinessProbe,
+			ImagePullPolicy: kapi.PullIfNotPresent,
+			VolumeMounts:    mounts,
+		},
+	}
 
-		if output {
-			list.Items, err = ocmdutil.ConvertItemsForDisplayFromDefaultCommand(cmd, list.Items)
-			if err != nil {
-				return err
-			}
+	if cfg.StatsPort > 0 && cfg.ExposeMetrics {
+		pc := generateMetricsExporterContainer(cfg, env)
+		if pc != nil {
+			containers = append(containers, *pc)
+		}
+	}
 
-			if err := f.PrintObject(cmd, list, out); err != nil {
-				return fmt.Errorf("Unable to print object: %v", err)
+	objects := []runtime.Object{}
+	for _, s := range secrets {
+		objects = append(objects, s)
+	}
+	if createServiceAccount {
+		objects = append(objects,
+			&kapi.ServiceAccount{ObjectMeta: kapi.ObjectMeta{Name: cfg.ServiceAccount}},
+			&authapi.ClusterRoleBinding{
+				ObjectMeta: kapi.ObjectMeta{Name: fmt.Sprintf("router-%s-role", cfg.Name)},
+				Subjects: []kapi.ObjectReference{
+					{
+						Kind:      "ServiceAccount",
+						Name:      cfg.ServiceAccount,
+						Namespace: namespace,
+					},
+				},
+				RoleRef: kapi.ObjectReference{
+					Kind: "ClusterRole",
+					Name: "system:router",
```

Comment:
I don't like constant references that aren't part of the API being taken
across package groups.  This has to be very deliberately changed.

On Wed, Feb 17, 2016 at 12:59 PM, Michail Kargakis <notifications@github.com
> wrote:

> In pkg/cmd/admin/router/router.go
> <https://github.com/openshift/origin/pull/7251#discussion_r53203936>:
>
> > +	}
> > +	if createServiceAccount {
> > +		objects = append(objects,
> > +			&kapi.ServiceAccount{ObjectMeta: kapi.ObjectMeta{Name: cfg.ServiceAccount}},
> > +			&authapi.ClusterRoleBinding{
> > +				ObjectMeta: kapi.ObjectMeta{Name: fmt.Sprintf("router-%s-role", cfg.Name)},
> > +				Subjects: []kapi.ObjectReference{
> > +					{
> > +						Kind:      "ServiceAccount",
> > +						Name:      cfg.ServiceAccount,
> > +						Namespace: namespace,
> > +					},
> > +				},
> > +				RoleRef: kapi.ObjectReference{
> > +					Kind: "ClusterRole",
> > +					Name: "system:router",
>
>
> https://github.com/openshift/origin/blob/d1a5a27d911e7836229992d128afa6d16dde30f4/pkg/cmd/server/bootstrappolicy/constants.go#L67
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7251/files#r53203936>.
>